Linear algebra: Expose the remaining classes listed in Eigen documentation (decompositions and solvers)

Lucas-Haubert opened the pull request:
This (draft) PR aims at adding the remaining classes in Eigen that we plan to use in practice. 

This PR is proposed jointly with the one in nanoeigenpy (https://github.com/Simple-Robotics/nanoeigenpy/pull/13).

What follows if the list of classes treated in this work: 
- [x] `<Eigen/LU>`: `FullPivLU`
- [x] `<Eigen/LU>`: `PartialPivLU`
- [x] `<Eigen/SVD>`: `JacobiSVD`
- [x] `<Eigen/SVD>`: `BDCSVD`
- [x] `<Eigen/Eigenvalues>`: `ComplexEigenSolver`
- [x] `<Eigen/Eigenvalues>`: `GeneralizedEigenSolver`
- [x] `<Eigen/Eigenvalues>`: `GeneralizedSelfAdjointSolver`
- [x] `<Eigen/Eigenvalues>`: `HessenbergDecomposition`
- [x] `<Eigen/Eigenvalues>`: `RealQZ`
- [x] `<Eigen/Eigenvalues>`: `RealSchur`
- [x] `<Eigen/Eigenvalues>`: `ComplexSchur`
- [x] `<Eigen/Eigenvalues>`: `Tridiagonalization`
- [x] `<Eigen/SparseLU>`: `SparseLU`
- [x] `<Eigen/SparseQR>`: `SparseQR`
- [x] `<Eigen/IterativeLinearSolvers>`: `BiCGSTAB`
- [x] `<Eigen/IterativeLinearSolvers>`: `IncompleteLUT`
- [x] `<Eigen/IterativeLinearSolvers>`: `IncompleteCholesky`

Lucas-Haubert committed:
decompositions: Expose classes

Lucas-Haubert committed:
readme: added lucas haubert

Lucas-Haubert committed:
decompositions: SVDBase, BDCSVD, JacobiSVD

Lucas-Haubert committed:
eigenvalues: Expose classes

Lucas-Haubert committed:
solvers: BiCGSTAB

Lucas-Haubert committed:
unittest: Comments about permutationP and permutationQ in FullPivLU

Lucas-Haubert committed:
unittest: PartialPivLU reconstructedMatrix to retrieve the original matrix

Lucas-Haubert committed:
Added comments and setThreshold default option in SVDBase

Lucas-Haubert committed:
solvers: Completed tests

Lucas-Haubert committed:
decompositions: comments and rewrites

Lucas-Haubert committed:
decompositions: Put ordering option AMD in SparseQR

Lucas-Haubert committed:
decompositions: Fixed computation options in BDCSVD

Lucas-Haubert committed:
decompositions: Rewrite comments in constructors in BDCSVD and JacobiSVD

Lucas-Haubert committed:
Set COLAMD Ordering in SParseQR

Lucas-Haubert committed:
decompositions: Completed tests

Lucas-Haubert committed:
Add IncompleteCholesky and IncompleteLUT

Lucas-Haubert committed:
Added all preconditioners in solvers

Lucas-Haubert committed:
Clean code: Set the same nomenclature of files for exposed classes and put MINRES in capital letters

Lucas-Haubert committed:
Set Eigen::Lower everywhere for UpLo template option

Lucas-Haubert committed:
Switch MINRES, IncompleteLUT and IncompleteCholesky in eigenpy.solvers scope

Lucas-Haubert committed:
Optional arguments with bp::optional

Lucas-Haubert committed:
README: Remove the 'SVD and QR to be added', but enounce them

Lucas-Haubert committed:
CI: Fix test_JacobiSVD

Lucas-Haubert committed:
is_approx: Put is_approx and set the precision

Lucas-Haubert committed:
Add const in some lambda functions on Solver& to stick to definition

Lucas-Haubert committed:
Add matrixQ in SparseQR

Lucas-Haubert committed:
Add actual date in copyrights

Lucas-Haubert committed:
Add methods matrixU and matrixL in SParseLU

Lucas-Haubert committed:
Capital letters in header names in sparse decompositions

jorisv commented on CMakeLists.txt:
```diff
@@ -227,6 +233,14 @@ set(${PROJECT_NAME}_DECOMPOSITIONS_HEADERS
+    include/eigenpy/decompositions/Tridiagonalization.hpp
+    include/eigenpy/decompositions/RealSchur.hpp
+    include/eigenpy/decompositions/ComplexEigenSolver.hpp
+    include/eigenpy/decompositions/ComplexSchur.hpp
```
```suggestion
    include/eigenpy/decompositions/ComplexSchur.hpp
    include/eigenpy/decompositions/FullPivLU.hpp
    include/eigenpy/decompositions/PartialPivLU.hpp
```

Lucas-Haubert replied:
Done.

jorisv commented on include/eigenpy/decompositions/FullPivLU.hpp:
```diff
@@ -0,0 +1,194 @@
+        .def("rcond", &Solver::rcond, bp::arg("self"),
+             "Returns an estimate of the reciprocal condition number of the "
+             "matrix.")
+#endif
```
We don't support eigen 3.3.0 anymore, you can remove this guard.

Lucas-Haubert replied:
Done.

jorisv commented on include/eigenpy/decompositions/PartialPivLU.hpp:
```diff
@@ -0,0 +1,140 @@
+        .def("rcond", &Solver::rcond, bp::arg("self"),
+             "Returns an estimate of the reciprocal condition number of the "
+             "matrix.")
+#endif
```
We don't support eigen 3.3.0 anymore, you can remove this guard.

jorisv commented on src/solvers/solvers.cpp:
```diff
@@ -13,23 +13,57 @@
+      expose("IdentityLeastSquaresConjugateGradient");
+  LeastSquaresConjugateGradientVisitor<DiagonalLeastSquaresConjugateGradient>::
+      expose("DiagonalLeastSquaresConjugateGradient");
 #endif
```
We don't support eigen 3.3.5 anymore, you can remove this guard.

jorisv commented on include/eigenpy/decompositions/sparse/SimplicialLDLT.hpp:
You should provide a back compatibility header.
You can check how is done here: https://github.com/stack-of-tasks/pinocchio/blob/devel/include/pinocchio/parsers/sample-models.hpp
Eigenpy provide the `EIGENPY_PRAGMA_DEPRECATED_HEADER` macro in fwd.hpp

jorisv commented on include/eigenpy/decompositions/sparse/SimplicialLLT.hpp:
You should provide a back compatibility header.
You can check how is done here: https://github.com/stack-of-tasks/pinocchio/blob/devel/include/pinocchio/parsers/sample-models.hpp
Eigenpy provide the `EIGENPY_PRAGMA_DEPRECATED_HEADER` macro in fwd.hpp

jorisv commented on include/eigenpy/decompositions/sparse/accelerate/Accelerate.hpp:
You should provide a back compatibility header.
You can check how is done here: https://github.com/stack-of-tasks/pinocchio/blob/devel/include/pinocchio/parsers/sample-models.hpp
Eigenpy provide the `EIGENPY_PRAGMA_DEPRECATED_HEADER` macro in fwd.hpp

jorisv commented on include/eigenpy/solvers/MINRES.hpp:
You should provide a back compatibility header.
You can check how is done here: https://github.com/stack-of-tasks/pinocchio/blob/devel/include/pinocchio/parsers/sample-models.hpp
Eigenpy provide the `EIGENPY_PRAGMA_DEPRECATED_HEADER` macro in fwd.hpp

jorisv commented on include/eigenpy/solvers/BiCGSTAB.hpp:
```diff
@@ -0,0 +1,41 @@
+ * Copyright 2025 INRIA
+ */
+
+#ifndef __eigenpy_bicgstab_hpp__
```
```suggestion
#ifndef __eigenpy_solvers_bicgstab_hpp__
```

jorisv commented on include/eigenpy/solvers/IncompleteCholesky.hpp:
```diff
@@ -0,0 +1,125 @@
+ * Copyright 2025 INRIA
+ */
+
+#ifndef __eigenpy_decompositions_sparse_incomplete_cholesky_hpp__
```
```suggestion
#ifndef __eigenpy_solvers_incomplete_cholesky_hpp__
```

jorisv commented on include/eigenpy/solvers/IncompleteLUT.hpp:
```diff
@@ -0,0 +1,108 @@
+ * Copyright 2025 INRIA
+ */
+
+#ifndef __eigenpy_decompositions_sparse_incomplete_lut_hpp__
```
```suggestion
#ifndef __eigenpy_solvers_incomplete_lut_hpp__
```

jorisv commented on src/solvers/solvers.cpp:
Like decomposition, to win compilation time and avoid big memory footprint it could be nice to split solver.cpp in many .cpp file.

jorisv commented on src/solvers/solvers.cpp:
```diff
@@ -13,23 +13,57 @@
+  BiCGSTABVisitor<BiCGSTAB<MatrixXd>>::expose("BiCGSTAB");
+  BiCGSTABVisitor<IdentityBiCGSTAB>::expose("IdentityBiCGSTAB");
+
+  MINRESSolverVisitor<MatrixXd>::expose("MINRES");
```
MINRES is now in the python eigenpy.solvers module. We must create an alias to still allow to call eigenpy.MINRES

jorisv commented on include/eigenpy/decompositions/sparse/SparseQR.hpp:
```diff
@@ -0,0 +1,226 @@
+            +[](QTransposeType& self, const VectorXd& other) -> VectorXd {
+              return VectorXd(self * other);
+            },
+            bp::args("self", "other"));
```
Can we convert the SparseQRMatrixQTransposeReturnTypeVisitor (and all other view) to a generic sparse matrix ?

Lucas-Haubert replied:
Yes, but even when the source code refers to the same intermediate class (as a return type) (`struct SparseQRMatrixQTransposeReturnType`) ?
For other views, they concern `SparseLU.hpp` and contain `solveInPlace` methods.

Lucas-Haubert replied:
Done, added `toSparse()` method in return type of `matrixQ()`.

Lucas-Haubert committed:
CMakeLists: Added FullPivLU and PartialPivLU headers

Lucas-Haubert committed:
Removed outpassed guards regarding eigen version in decompositions and solvers

Lucas-Haubert committed:
Include guards: Set name that corresopnds to directory

Lucas-Haubert committed:
Include guards: Updated date in copyright

Lucas-Haubert committed:
MINRES: Backward compatibility to call from eigenpy and eigenpy.solvers

Lucas-Haubert committed:
Split solvers.cpp into separated .cpp as in decompositions

Lucas-Haubert committed:
fwd:hpp: Corrected definition of macro EIGENPY_PRAGMA_DEPRECATED_HEADER following pinocchio

Lucas-Haubert committed:
Add back compatibility headers due to changes in name and directory of headers

Lucas-Haubert committed:
unittest: Simplify test_GeneralizedEigenSolver to reduce computation time

Lucas-Haubert committed:
Back compatibility accelerate.hpp

Lucas-Haubert committed:
unittest: Reduced computation time of tests

Lucas-Haubert committed:
CHANGELOG: Update

Lucas-Haubert committed:
SparseQR: Add toSparse method in matrixQ return type to extract the related Eigen SparseMatrix

nim65s committed:
nix: patch eigen

Nixpkgs uses eigen v3.4.0-unstable-2022-04-19,
aka https://gitlab.com/libeigen/eigen/-/merge_requests/974,
because that is required by onnxruntime.
Ref. https://github.com/NixOS/nixpkgs/pull/364362

But that fix actually break other things, which are then fixed in
https://gitlab.com/libeigen/eigen/-/merge_requests/977

So we need to apply that here, and hope for the next eigen version
